File: 백준/색종이만들기2630.cpp
#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;
int a[150][150];
int white = 0;
int blue = 0;
void dfs(int x,int y,int n)
{
    bool flag = true;
    for(int i=x;i<x+n;i++)
    {
        for(int j=y;j<y+n;j++)
        {
            if(a[i][j]!=a[x][y])
            {
                flag = false;
                break;
            }
        }
    }
    if(flag==false)
    {
        dfs(x,y,n/2);
        dfs(x,y+n/2,n/2);
        dfs(x+n/2,y,n/2);
        dfs(x+n/2,y+n/2,n/2);
    }
    else
    {
        if(a[x][y]==1)
        {
            blue+=1;
        }
        else
        {
            white+=1;
        }
    }
}
int main()
{
    int n;
    cin >> n;
    for(int i=0;i<n;i++)
    {
        for(int j=0;j<n;j++)
        {
            cin >> a[i][j];
        }
    }
    dfs(0,0,n);
    cout << white << '\n';
    cout << blue << '\n';
    return 0;
}

File: Java/12.분할정복/색종이만들기2630.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Collections;
import java.util.Comparator;
import java.util.LinkedList;
import java.util.List;
import java.util.StringTokenizer;

public class Main {

	static int[][] board = new int[150][150];
	static int white = 0;
	static int blue = 0;

	static void dfs(int px, int py, int n) {
		boolean flag = true;
		for(int i=0;i<n;i++) {
			for(int j=0;j<n;j++) {
				if(board[px+i][py+j]!=board[px][py]) {
					flag = false;
				}
			}
		}
		if(flag==false) {
			dfs(px,py,n/2);
			dfs(px+n/2,py,n/2);
			dfs(px+n/2,py+n/2,n/2);
			dfs(px,py+n/2,n/2);
		} else {
			if(board[px][py]==1) blue+=1;
			else white+=1;
			return;
		}
	}

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		StringTokenizer st;

		int n = Integer.parseInt(br.readLine());
		for(int i=0;i<n;i++) {
			st = new StringTokenizer(br.readLine());
			for(int j=0;j<n;j++) {
				board[i][j] = Integer.parseInt(st.nextToken());
			}
		}
		dfs(0,0,n);
		System.out.println(sb.append(white+"\n"+blue));
	}
}
